LibGit2: build libgit2 with libssh2 support

wildart opened the pull request:
Build `libgit2` with `libssh2` support
- Linux: SSL: OpenSSL,  SSH: libssh2 +mbedTLS
- macOS: SSL: Native, SSH: libssh2 + mbedTLS
- Windows: SSL: Native, SSH: libssh2 +Native cryptolib

Fixes https://github.com/JuliaLang/julia/issues/16041, see also https://github.com/JuliaLang/julia/pull/17246

@Keno @tkelman

wildart committed:
build libgit2 with libssh2 support
- Linux: OpenSSL & mbedTLS cryptolib for libssh2
- macOS: Native SSL & mbedTLS cryptolib for libssh2
- Windows: Native SSL & cryptolib

tkelman commented on .travis.yml:
```diff
@@ -18,7 +20,7 @@ matrix:
             - g++-5-multilib
             - make:i386
-            - libssl-dev:i386
+            - libssl-dev:i386s
```
typo?

Keno commented:
Love it! Thanks @wildart.

tkelman commented on deps/libssh2.mk:
```diff
@@ -0,0 +1,56 @@
+## libssh2
+
+LIBSSH2_GIT_URL := git://github.com/wildart/libssh2.git
```
can we structure this as a patch file instead of using your repo?

wildart replied:
ok

wildart committed:
fix typo

tkelman commented on deps/patches/mbedtls.patch:
```diff
@@ -0,0 +1,27 @@
+diff --git include/mbedtls/config.h include/mbedtls/config.h
```
can we call this file something a little more specific? `mbedtls-config.patch` or similar?

tkelman commented:
Great, this does look promising. We'll need to mention libssh and mbedtls in the readme and license along with the other deps.

wildart committed:
get libssh2 from official repo & patch it

wildart committed:
rename mbedtls patch

tkelman commented on deps/libssh2.version:
```diff
@@ -1,2 +1,2 @@
-LIBSSH2_BRANCH=mbedtls
-LIBSSH2_SHA1=a486c43e4a699d319a5afdb139545843769bcc0d
+LIBSSH2_BRANCH=master
```
should use a release tag name if possible

wildart commented:
For some reason `mbedtls` linking fails on CI services. Any thoughts?

Keno commented:
Can reproduce locally by doing `-j8`. Missing dependency somewhere?

tkelman commented on deps/libssh2.version:
```diff
@@ -0,0 +1,2 @@
+LIBSSH2_BRANCH=master
+LIBSSH2_SHA1=7934c9ce2a029c43e3642a492d3b9e494d1542be
```
can we use https://github.com/libssh2/libssh2/releases/tag/libssh2-1.7.0 or is there something on master we need?

wildart replied:
nope, I developed mdebtls support on master

tkelman replied:
can the patch be rebased and still work?

wildart replied:
i'll look at it

tkelman replied:
I can also probably take a look at rebasing it but I'm not sure what to test

wildart commented:
mdebtls does not have any dependency.

wildart committed:
use release tag for libssh2

tkelman commented on deps/mbedtls.mk:
```diff
@@ -15,7 +15,7 @@ MBEDTLS_OBJ_TARGET := $(build_shlibdir)/libmbedtls.$(SHLIB_EXT) \
 
 MBEDTLS_OPTS := $(CMAKE_COMMON) -DUSE_SHARED_MBEDTLS_LIBRARY=ON \
-		-DENABLE_PROGRAMS=OFF -DCMAKE_BUILD_TYPE=Release \
+		-DENABLE_PROGRAMS=OFF -DENABLE_TESTING=OFF -DCMAKE_BUILD_TYPE=Release \
```
would be ideal if we can get these to work, but I guess that can be done during the RC period if it isn't working yet

tkelman commented on deps/mbedtls.mk:
```diff
@@ -49,7 +49,7 @@ $(MBEDTLS_OBJ_SOURCE): $(BUILDDIR)/mbedtls-$(MBEDTLS_VER)/Makefile
 $(BUILDDIR)/mbedtls-$(MBEDTLS_VER)/checked: $(MBEDTLS_OBJ_SOURCE)
 ifeq ($(OS),$(BUILD_OS))
-	$(MAKE) -C $(dir $@) test
+	-$(MAKE) -C $(dir $@) test
```
I don't think this rule gets run by default unless you specifically request it, so please don't prefix it with `-`

wildart replied:
I'm testing mbedtls build: test programs where linked with errors

tkelman replied:
so we should see the errors instead of skipping them if you manually run `make -C deps test-mbedtls`, no?

tkelman commented on deps/patches/libssh2-mbedtls.patch:
```diff
@@ -35,6 +35,47 @@ index 9d6c0be..8bc1292 100644
+ the revised BSD license.
+ 
+-Web site: http://www.libssh2.org/
++Web site: https://www.libssh2.org/
```
we should really make this patch as small as it can be and still work, but that's "after it works" cleanup we can do during RC's

wildart replied:
hopefully it will be merged into libssh2 master

tkelman replied:
Yes hopefully it will, we haven't heard much there though right? If I had to guess, maybe the reason I wasn't able to get it to pass tests was because I was trying to use stock unmodified mbedtls instead of applying the config patch?

wildart replied:
yes, by default, mbedtls internal buffers for bignum manipulations are small

wildart committed:
fix mbedtls build

tkelman commented:
If / when this is ready, let me know and I'll trigger a buildbot run off the branch to see if it works there too.

wildart commented:
@tkelman Travis directory caching screws this PR build. It looks like x86 and x86_64 builds use the same `$TRAVIS_BUILD_DIR/deps/build` and `$TRAVIS_BUILD_DIR/deps/srccache` folders

wildart commented:
DO NOT MERGE

tkelman commented:
you crossed it out, but 32 and 64 bit use different caches because they set different values for `compiler`. We may need to occasionally clear the cache for this pr if it has uploaded its own, to make it go back to using the master cache. Please don't delete the master cache though, that would take an hour or two on travis to rebuild.

wildart committed:
fix libgit2 & libssh2 build

wildart commented:
@tkelman I fixed it. You could try to run in through a buildbot

tkelman commented on .travis.yml:
```diff
@@ -79,7 +83,7 @@ before_install:
         BUILDOPTS="$BUILDOPTS VERBOSE=1 USE_BLAS64=0 SUITESPARSE_INC=-I$(brew --prefix suite-sparse-julia)/include FORCE_ASSERTIONS=1 STAGE2_DEPS=utf8proc";
         BUILDOPTS="$BUILDOPTS LIBBLAS=-lopenblas LIBBLASNAME=libopenblas LIBLAPACK=-lopenblas LIBLAPACKNAME=libopenblas";
-        for lib in LLVM SUITESPARSE ARPACK BLAS FFTW LAPACK GMP MPFR PCRE LIBUNWIND LIBGIT2; do
+        for lib in LLVM SUITESPARSE ARPACK BLAS FFTW LAPACK GMP MPFR PCRE LIBUNWIND; do
```
does the homebrew libgit2 get built with or without ssh support?

wildart replied:
we going to use our `libssh2` based on mbedTLS cryptolib, so it is necessary to rebuild `libgit2` as well

tkelman replied:
yeah that makes sense, but we might want to do so in homebrew-juliadeps to save the travis build time of recompiling all 3 libraries from source on every build

tkelman committed:
Minor grammar fix in a comment

and revert some unnecessary whitespace additions to make
the pr diff smaller

tkelman commented on deps/libgit2.mk:
```diff
@@ -20,26 +20,36 @@ LIBGIT2_OPTS += -DBUILD_CLAR=OFF -DDLLTOOL=`which $(CROSS_COMPILE)dlltool`
 	echo 1 > $@
+
 $(LIBGIT2_OBJ_TARGET): $(LIBGIT2_OBJ_SOURCE) | $(build_shlibdir)
+ifeq ($(BUILD_OS),WINNT)
```
what is needed from make install other than the shared library on linux and mac, and why? if only the dll is needed on windows, should this also apply for cross-compile and therefore be `OS` instead of `BUILD_OS` ?

wildart replied:
this is an issue with msys staged install (something related to our make-install or cmake - files are copied into wrong folder). Cygwin build doesn't have this issue.

tkelman replied:
ok thanks for confirming - which wrong folder, something like /lib instead of /bin for dll's? one more reason to just drop msys2 support until we actually have a good reason to worry about it

wildart replied:
right,  /lib instead of /bin

tkelman commented on deps/mbedtls.mk:
```diff
@@ -0,0 +1,74 @@
+ifeq ($(OS),WINNT)
+MBEDTLS_OPTS += -DENABLE_ZLIB_SUPPORT=OFF
+ifeq ($(BUILD_OS),WINNT)
+MBEDTLS_OPTS += -G"MSYS Makefiles" -DENABLE_TESTING=OFF
```
why was it necessary to disable testing?

wildart replied:
test are not working on windows MSYS build (incorrect paths)

tkelman replied:
runtime failure or build time failure?

tkelman commented on deps/mbedtls.mk:
```diff
@@ -0,0 +1,74 @@
+MBEDTLS_OPTS += -G"MSYS Makefiles" -DENABLE_TESTING=OFF
+endif
+else
+MBEDTLS_OPTS += -DENABLE_ZLIB_SUPPORT=ON
```
do we absolutely need this? if it results in the binaries being linked against zlib, I don't think we're currently including those in our linux binaries which need to be standalone

wildart replied:
ok, I'll disable it.

tkelman replied:
also cc @malmaud do you know whether this makes a difference for MbedTLS.jl's usage of the library?

wildart replied:
Make sure that `libz` is not installed on the machine, because `libgit2` could link to it as well.

tkelman replied:
can we explicitly disable that?

wildart replied:
it's need to compress for data during http(s) requests

tkelman replied:
is that a major noticeable performance difference when cloning https repos? I suspect the majority of systems will have zlib installed, but it's possible some minimal containers might not. Let me test in a bare docker container and see what happens.

tkelman commented on deps/mbedtls.mk:
```diff
@@ -0,0 +1,74 @@
+distclean-mbedtls:
+	-rm -rf $(SRCDIR)/srccache/$(MBEDTLS_SRC).tgz
+	-rm -rf $(SRCDIR)/srccache/$(MBEDTLS_SRC)
+	-rm -rf $(BUILDDIR)/mbedtls-$(MBEDTLS_VER)
```
`distclean-mbedtls` should just depend on `clean-mbedtls`

tkelman commented on deps/libssh2.mk:
```diff
@@ -0,0 +1,57 @@
+ifeq ($(OS),WINNT)
+	$(MAKE) -C $(BUILDDIR)/$(LIBSSH2_SRC_DIR) install
+else
+	$(call make-install,$(LIBSSH2_SRC_DIR),)
```
what does this do differently from the above line?

wildart replied:
winnt - direct install, otherwise staged

tkelman replied:
is staged install broken then? on both cygwin-cross and msys2?

wildart replied:
sorry, don't remember

tkelman replied:
Ok. Thanks for being up late on a friday working through this. I'll try using the staged `make-install` and see if it works in cygwin for me, if so I'll change it to just BUILD_OS under the assumption that it might have just been an MSYS2 issue.

tkelman commented on deps/patches/libgit2-custom-tls.patch:
```diff
@@ -0,0 +1,65 @@
+ #ifdef GIT_THREADS
+ 		| GIT_FEATURE_THREADS
+ #endif
+-#if defined(GIT_OPENSSL) || defined(GIT_WINHTTP) || defined(GIT_SECURE_TRANSPORT)
```
should we propose a patch that adds a configurable custom option to trigger this?

wildart replied:
https://github.com/libgit2/libgit2/pull/3850

tkelman commented:
This is getting very close. Here are the binaries, they need testing:

https://julianightlies.s3.amazonaws.com/bin/linux/x64/0.5/julia-0.5.0-ed41e46e31-linux64.tar.gz
https://julianightlies.s3.amazonaws.com/bin/linux/x86/0.5/julia-0.5.0-ed41e46e31-linux32.tar.gz
https://julianightlies.s3.amazonaws.com/bin/osx/x64/0.5/julia-0.5.0-ed41e46e31-osx.dmg
https://julianightlies.s3.amazonaws.com/bin/winnt/x64/0.5/julia-0.5.0-ed41e46e31-win64.exe
https://julianightlies.s3.amazonaws.com/bin/winnt/x86/0.5/julia-0.5.0-ed41e46e31-win32.exe

I'm going to push a commit here that takes unrelated changes out of the libssh2 patch (and a few other relatively trivial things).

tkelman commented on deps/libgit2.mk:
```diff
@@ -20,26 +20,36 @@ LIBGIT2_OPTS += -DBUILD_CLAR=OFF -DDLLTOOL=`which $(CROSS_COMPILE)dlltool`
+ifeq ($(OS),Linux)
 	-cd $(SRCDIR)/srccache/$(LIBGIT2_SRC_DIR) && patch -p1 -f < $(SRCDIR)/patches/libgit2-require-openssl.patch
+endif
+	-cd $(SRCDIR)/srccache/$(LIBGIT2_SRC_DIR) && patch -p0 -f < $(SRCDIR)/patches/libgit2-ssh.patch
```
I assume this should be `libgit2-custom-tls.patch` ?

wildart replied:
nope, libgit2-custom-tls.patch is for openssl-less libgit2 build

tkelman replied:
are we doing that here, or should that patch be moved to the other, more WIP PR?

wildart replied:
I'll remove it for now

tkelman committed:
Remove many unrelated changes from libssh2-mbedtls.patch

tkelman commented:
I can never get ssh to work with git so don't personally use it and can't easily test with it here (testing help appreciated), but at least the existing functionality seems to work even in a minimal docker container so maybe we don't need to worry about the zlib thing on Linux.

I'll add the readme changes. If you had to guess, what do you think the minimum versions of libssh and mbedtls would be? In case some distro were to try building against something earlier? Of course if you're using system libssh you wouldn't need to use mbedtls, but maybe you could try to use our source build of libssh against an older system mbedtls?

tkelman committed:
Add libssh2 and mbedtls to LICENSE and README

add 1.7 and 2.2 as minimum libssh2 and mbedtls versions, respectively

tkelman committed:
Remove libgit2-custom-tls.patch for now

not being used yet in this PR

tkelman committed:
Minor makefile cleanups

make cleaning and installation a bit more uniform

Move CMAKE_PREFIX_PATH to CMAKE_COMMON

tkelman commented on deps/libgit2.mk:
```diff
@@ -7,9 +7,9 @@ $(eval $(call git-external,libgit2,LIBGIT2,CMakeLists.txt,build/libgit2.$(SHLIB_
 LIBGIT2_OBJ_TARGET := $(build_shlibdir)/libgit2.$(SHLIB_EXT)
 
-LIBGIT2_OPTS := $(CMAKE_COMMON) -DTHREADSAFE=ON
+LIBGIT2_OPTS := $(CMAKE_COMMON) -DCMAKE_BUILD_TYPE=Release -DTHREADSAFE=ON -DCMAKE_PREFIX_PATH=$(build_prefix)
```
I wonder whether we should add `CMAKE_PREFIX_PATH` to `CMAKE_COMMON`

wildart replied:
yes

wildart commented:
libssh 1.7.0, mbedtls 2.2.1, libgit2 0.24 - we need to build them all, because there is no guaranty that system libraries will provide required feature set.

tkelman commented:
what's the required feature set that wouldn't be in libssh 1.6 or mbedtls 2.1?

oops, reason to care about msys2 = appveyor, will fix
```
CMake Error at tests/CMakeLists.txt:116 (message):
  Could not create symbolic link for:
  C:/projects/julia/deps/srccache/mbedtls-2.2.1-gpl/tests/data_files -->
  Invalid switch - "projects".
-- Configuring incomplete, errors occurred!
See also "C:/projects/julia/deps/build/mbedtls-2.2.1/CMakeFiles/CMakeOutput.log".
/c/projects/julia/deps/mbedtls.mk:36: recipe for target 'build/mbedtls-2.2.1/Makefile' failed
make[1]: *** [build/mbedtls-2.2.1/Makefile] Error 1
make[1]: *** Waiting for unfinished jobs....
```

tkelman commented on deps/mbedtls.mk:
```diff
@@ -7,12 +7,8 @@ else
-			$(build_shlibdir)/libmbedx509.$(SHLIB_EXT) \
-			$(build_shlibdir)/libmbedcrypto.$(SHLIB_EXT)
+MBEDTLS_OBJ_SOURCE := $(BUILDDIR)/mbedtls-$(MBEDTLS_VER)/library/libmbedcrypto.$(SHLIB_EXT)
+MBEDTLS_OBJ_TARGET := $(build_shlibdir)/libmbedcrypto.$(SHLIB_EXT)
```
just for clarification purposes, is this the only one of the three that libssh2 needs, or does it actually need all 3?

tkelman replied:
ping. not critical, but for the record would be good to have answers written down to these last minor clarifications while it's fresh in memory

wildart replied:
I left only one, because if I put three and use this variable as dependency in other Makefiles (libssh2) then make builds it 3 times.

tkelman commented:
As long as my latest little tweaks are green on CI, I think this is done. If someone can independently confirm that using a package over ssh works with this branch or the binaries posted above, I think we can merge.

tkelman commented on deps/patches/libgit2-ssh.patch:
```diff
@@ -0,0 +1,23 @@
+ # Optional external dependency: libssh2
+ IF (USE_SSH)
+-	PKG_CHECK_MODULES(LIBSSH2 libssh2)
++	FIND_PACKAGE(Libssh2)
```
what was the story and/or link on whether upstream needs or wants this patch?

tkelman replied:
ping?

wildart replied:
same thing -  MSYS2, pkg-config incorrectly reports installation location of libssh2

tkelman replied:
thanks. if this is yet another msys2 specific issue I'm going to have to migrate appveyor over to cross building from cygwin.

davidanthoff commented:
I just tried this with julia-0.5.0-90c296dbb9-win64.exe from above. This is what I get:
````
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.5.0-dev+5445 (2016-07-16 06:59 UTC)
 _/ |\__'_|_|_|\__'_|  |  master/90c296d (fork: 12 commits, 2 days)
|__/                   |  x86_64-w64-mingw32

julia> Pkg.clone("git@github.com:davidanthoff/Judyp.jl.git")
INFO: Cloning Judyp from git@github.com:davidanthoff/Judyp.jl.git
Private key location for 'git@github.com' [C:\Users\david\.ssh\id_rsa]:
Passphrase for C:\Users\david\.ssh\id_rsa:INFO: Computing changes...
INFO: No packages to install, update or remove

julia> Pkg.update()
INFO: Updating METADATA...
INFO: Computing changes...
INFO: No packages to install, update or remove
ERROR: Update finished with errors.
=> Package Judyp cannot be updated.
GitError(Code:ERROR, Class:SSH, error authenticating: failed connecting agent)
 in macro expansion at .\libgit2\error.jl:98 [inlined]
 in #fetch#52(::Base.LibGit2.FetchOptions, ::String, ::Function, ::Base.LibGit2.GitRemote, ::Array{AbstractString,1}) at .\libgit2\remote.jl:70
 in (::Base.LibGit2.#kw##fetch)(::Array{Any,1}, ::Base.LibGit2.#fetch, ::Base.LibGit2.GitRemote, ::Array{AbstractString,1}) at .\<missing>:0
 in #fetch#91(::String, ::String, ::Array{AbstractString,1}, ::Nullable{Base.LibGit2.CachedCredentials}, ::Function, ::Base.LibGit2.GitRepo) at .\libgit2\libgit2.jl:159
 in (::Base.LibGit2.#kw##fetch)(::Array{Any,1}, ::Base.LibGit2.#fetch, ::Base.LibGit2.GitRepo) at .\<missing>:0
 in (::Base.Pkg.Entry.##39#45)(::Base.LibGit2.GitRepo) at .\pkg\entry.jl:425
 in with(::Base.Pkg.Entry.##39#45, ::Base.LibGit2.GitRepo) at .\libgit2\types.jl:660
 in update(::String, ::Set{String}) at .\pkg\entry.jl:415
 in (::Base.Pkg.Dir.##2#3{Array{Any,1},Base.Pkg.Entry.#update,Tuple{String,Set{String}}})() at .\pkg\dir.jl:31
 in cd(::Base.Pkg.Dir.##2#3{Array{Any,1},Base.Pkg.Entry.#update,Tuple{String,Set{String}}}, ::String) at .\file.jl:48
 in #cd#1(::Array{Any,1}, ::Function, ::Function, ::String, ::Vararg{Any,N}) at .\pkg\dir.jl:31
 in update() at .\pkg\pkg.jl:210
 in eval(::Module, ::Any) at .\boot.jl:234
 in eval_user_input(::Any, ::Base.REPL.REPLBackend) at .\REPL.jl:62
 in macro expansion at .\REPL.jl:92 [inlined]
 in (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at .\event.jl:46
 in update(::String, ::Set{String}) at .\pkg\entry.jl:459
 in (::Base.Pkg.Dir.##2#3{Array{Any,1},Base.Pkg.Entry.#update,Tuple{String,Set{String}}})() at .\pkg\dir.jl:31
 in cd(::Base.Pkg.Dir.##2#3{Array{Any,1},Base.Pkg.Entry.#update,Tuple{String,Set{String}}}, ::String) at .\file.jl:48
 in #cd#1(::Array{Any,1}, ::Function, ::Function, ::String, ::Vararg{Any,N}) at .\pkg\dir.jl:31
 in update() at .\pkg\pkg.jl:210
 in eval(::Module, ::Any) at .\boot.jl:234
 in macro expansion at .\REPL.jl:92 [inlined]
 in (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at .\event.jl:46

julia>
````
Apart from the error, it also shouldn't ask for a passphrase, my ssh key doesn't have one. And is the private key location cached somewhere?

And there should be a line break after I enter the passphrase and the next log message (i.e. the line ``Passphrase for C:\Users\david\.ssh\id_rsa:INFO: Computing changes...`` should be split in two).

I didn't test anything else yet.

Keno commented:
This basically works for me, except that instead of prompting for my ssh key, it should respect `SSH_AUTH_SOCK` and let the operating system handle it.

Keno commented:
I'll try to make the agent thing work.

tkelman commented:
what's better, merging and revising it on master, or tweaking inside this pr?

Keno commented:
Merging and revisiting on master probably. I do consider not having ssh agent support blocking for the release though, since otherwise you have to enter your password every time.

davidanthoff commented:
This PR doesn't break anything, i.e. SSH doesn't work on master in any case, right? Then I'd vote for merging and then fixing, is probably easier for people to try out in that case because they can just use the nightly build etc.

tkelman commented:
though now it counts as a bugfix and not a new feature so we aren't blocked from feature freeze

davidanthoff commented:
@Keno, plus https also doesn't work, minimally on windows (#17423, #17422), so at this moment there is no way to use private repos, regardless of how often one is willing to type in passwords.

tkelman commented:
Kind of a bummer that the mbedtls config patch makes its tests fail (`mpi-suite` and `rsa-suite`), and the libssh2 tests fail without it. Would we need to patch mbedtls' tests too to account for the changes?
